lambda/src/main/scala/pricemigrationengine/migrations/Newspaper2024Migration/Newspaper2024MigrationEstimation.scala

```diff
@@ -0,0 +1,253 @@
+
+  // RatePlanDetails was introduced to help properly test the data gathering to build the PriceData
+  // It turned out to be particularly useful for testing that the logic was correct
+  // It is currently limited to Newspaper2024Migration, but could be generalised to other (future) migrations
+  case class SubscriptionData2024(
+      productName: String,
+      ratePlan: ZuoraRatePlan,
+      ratePlanName: String,
+      billingPeriod: BillingPeriod,
+      currency: String,
+      currentPrice: BigDecimal,
+      targetRatePlanId: String,
+  )
+
+  // We have an unusual scheduling for this migration and Newspaper2024BatchId is used to
+  // decide the correct start date for each subscription.
+  sealed trait Newspaper2024BatchId
+  object MonthliesPart1 extends Newspaper2024BatchId // First batch of monthlies
+  object MonthliesPart2 extends Newspaper2024BatchId // Second batch of monthlies
+  object MoreThanMonthlies extends Newspaper2024BatchId // Quarterlies, Semi-Annuals and Annuals
+
+  def subscriptionToMigrationProductName(subscription: ZuoraSubscription): Either[String, String] = {
+    // We are doing a multi product migration. This function tries and retrieve the correct product given a
+    // subscription.
+    val migrationProductNames = List("Newspaper Delivery", "Newspaper Digital Voucher", "Newspaper Voucher")
+    val names = subscription.ratePlans
+      .filter(ratePlan => ratePlan.lastChangeType == None || ratePlan.lastChangeType == Some("Add"))
+      .map(ratePlan => ratePlan.productName)
+      .filter(name => migrationProductNames.contains(name))
+      .distinct
+    names match {
+      case Nil =>
+        Left(
+          s"[error: d5fb6922] could not determine migration product name for subscription ${subscription.subscriptionNumber}; no name to choose from"
+        )
+      case name :: Nil => Right(name)
+      case _ =>
+        Left(
+          s"[error: d3ecd18d] could not determine migration product name for subscription ${subscription.subscriptionNumber}; more than one name to choose from"
+        )
+    }
+  }
+
+  def ratePlanToBillingPeriod(ratePlan: ZuoraRatePlan): Option[BillingPeriod] = {
+    for {
+      ratePlanCharge <- ratePlan.ratePlanCharges.headOption
+      billingPeriod <- ratePlanCharge.billingPeriod
+    } yield BillingPeriod.fromString(billingPeriod)
+  }
+
+  def ratePlanToCurrency(ratePlan: ZuoraRatePlan): Option[String] = {
+    for {
+      ratePlanCharge <- ratePlan.ratePlanCharges.headOption
+    } yield ratePlanCharge.currency
+  }
+
+  def subscriptionToSubscriptionData2024(subscription: ZuoraSubscription): Either[String, SubscriptionData2024] = {
+    val productName = subscriptionToMigrationProductName(subscription).toOption.get
+    val ratePlans = {
+      subscription.ratePlans
+        .filter(ratePlan => ratePlan.productName == productName)
+        .filter(ratePlan => ratePlan.lastChangeType == None || ratePlan.lastChangeType == Some("Add"))
```

Comment:
Could extract this condition as it's repeated above

Reply:
The two side by side look like this. 

```
val names = subscription.ratePlans
  .filter(ratePlan => ratePlan.lastChangeType == None || ratePlan.lastChangeType == Some("Add"))
  .map(ratePlan => ratePlan.productName)
  .filter(name => migrationProductNames.contains(name))
  .distinct

val ratePlans = {
  subscription.ratePlans
    .filter(ratePlan => ratePlan.productName == productName)
    .filter(ratePlan => ratePlan.lastChangeType == None || ratePlan.lastChangeType == Some("Add"))
    .distinct
```

I could put `ratePlan.lastChangeType == None || ratePlan.lastChangeType == Some("Add")` into its own function if you feel strongly about it :)

Reply:
Never mind. It's actually nicer to abstract it: https://github.com/guardian/price-migration-engine/pull/964/commits/aeac2a052d47ca274bd06827e53d08003eb1aeb8